### ISSUE TITLE
Ideas for synchronous initialisation

### ISSUE THREAD

**Issue by tamlyn** [dmfay/massive-js#602]
I love Massive but there's one aspect that I find hard to deal with is the fact that initialisation is asynchronous.

With `node-pg` or `pg-promise` I'm used to having a file in which I create the database connection then export the connection object. I can then import that file as `db` in the rest of my app and I don't have to think about it. With `massive` I can't do that because the result of the connection is a promise that must be awaited before I can call any functions on it. Even though the app will run for days or weeks, every call to the database must await a promise which was resolved in the first few milliseconds after startup.

### Use proxies

Of course, the connecting to the database is always asynchronous but `node-pg` hides this fact by buffering any queries that come is before the connection is established. Massive can't currently do this because it doesn't know its own API until introspection has completed. 

Unless we use a proxy object to capture all method calls on the massive object and then either run the query or throw a type error once introspection has finished.

Proxies are supported in [Node >6.4](https://node.green/#ES2015-built-ins-Proxy)

### Cached introspection

Another possibility would be to compute the result of introspection at build time and store it in a file. This could then be read synchronously on startup.

### Hybrid promise

Or the promise returned on connection could be extended with the results of the introspection to *become* the Massive object. This would mean any queries made before introspection completes would fail but in practice these are known in advance so can be accounted for by awaiting the promise. I actually tried this approach on a previous project that used Massive and it works. But it does have some edge cases around reconnection when running tests.

### Other ideas?

Or perhaps I'm missing something which means this isn't even a problem?

**Comment by dmfay** [comment]
Similar questions have come up before, most recently in #590; see if my answer there works for you. I don't think there's really any _good_ way to implement around introspection being "hard-async", if you will; there are various kludges, but all of them come with a price. Proxies are slow, cached results may not be valid when it actually starts up for real, with your "hybrid" you've (as you said) still got to await it _anyway_ to be sure it's usable, [deasync](https://www.npmjs.com/package/deasync) is a whole other dependency with its own set of issues and does a _lot_ of mucking about in the Node engine.

So far I honestly haven't thought the init step is that big a deal, since it's a one-time cost, fairly easily abstracted, and init functions in the frameworks I've used have dealt with asynchronicity well enough. Async is Node's bread and butter and it's been getting easier and easier to work with.

**Comment by tamlyn** [comment]
Ah, I hadn't realised there were performance implications to using proxies. Is that the only issue you foresee with that approach? It seems likely that the performance penalty would be minimal when compared to the speed of the database query as a whole. 

According to [this post about proxy performance](http://thecodebarbarian.com/thoughts-on-es6-proxies-performance), a single proxy invocation is ~0.18µs slower than a vanilla function invocation. A 180ms database query is 1,000,000 times slower than the performance you just lost to using a proxy. I'm willing to give that up.

And if we're on the subject of performance, what about the cost of awaiting the connection each time? That means creating a promise and queueing up a new microtask each time. I'd put money on that taking much more than 0.18μs.

Of course even if you did want to go down this route, it might have all kinds of other pitfalls! I'd rather find a solution that works now. Do you have any examples of (open source) apps using massive that I could look at for inspiration on how to structure the database access logic?

**Comment by dmfay** [comment]
The added complexity (and supporting/maintaining that complexity) is the other aspect that gives me pause about proxies. But let's back up a bit -- I'm still not getting why awaiting a promise on initialization is a significant annoyance. The model I've used with koa and Express is [in the docs](https://dmfay.github.io/massive-js/#koa-example). What does yours look like now and what are you shooting for?

**Comment by vitaly-t** [comment]
My two cents on the caching idea.

Caching can only be efficient when it is implemented according to the business logic of the application. A general-purpose database library cannot encapsulate caching efficiently.

**Comment by dmfay** [comment]
I don't think it was about performance, it's that it can add complexity when you have to acquire the instance inside your promise handling instead of outside it.
